### ISSUE TITLE
rust-analyzer often fails to update its file mapping when moving between self-named module files and mod.rs

### ISSUE THREAD

**Issue by svix-jplatte** [rust-lang/rust-analyzer#19671]
It happens somewhat regularily for me that I switch between `foo.rs` and `foo/mod.rs` during refactoring, and most of the times I do this, RA seems to fail to update its internal module to file path mapping. Trying to reproduce this in a small sample project showed me that it doesn't happen all the time, so I can not provide exact reproduction steps but I was able to trigger it with only two modules after a couple of renames back and forth (the one that triggered it happened from within VSCode: I moved the file first so it was `foo/bar/bar.rs`, then renamed it to `mod.rs`).

When this happens, rust-analyzer doesn't generate any errors for the `mod` statement, but Ctrl+Clicking on it tries to load the wrong file. When opening the renamed module file, there's an unlinked file diagnostic at the start of the file and barely any RA functionality works within it, just like for other unlinked files. The issue is only fixable by restarting the server (or the entire extension process / VSCode window), 'reload workspace' doesn't help.

**rust-analyzer version**: 0.4.2435-standalone
**rustc version**: 1.88.0-nightly (d5b4c2e4f 2025-04-02)
**editor or extension**: VSCode, extension version 0.4.2435 (pre-release, from Open VSX)
**relevant settings**:
```json5
{
    "rust-analyzer.files.watcher": "server",
}
```

**Comment by UYousafzai** [comment]
specially on remote connections, git cloned a repository and Zed failed to recognise the folders pulled, additionally there is no manual refresh which I feel like there should be regardless of its auto detect feature, currently I am stuck on refreshing via reloading the whole folder.
